### ISSUE TITLE
Paste auto indentation, Inconsistent json tab width,  keyboard code visibility

### ISSUE THREAD

**Issue by dhaigit** [tyron12233/CodeAssist#186]
1. Text pasted inside a pair of "{}" does not get proper auto indentation or format.

For example, try to cut the j-for-loop and paste it inside the i-for-loop.

for (int i = 0; i < 10; i++) {

}
for (int j = 0; j < 10; j++) {

}

2. Auto format (from "..." menu) forms 2-space indentations while a tab character is 4 spaces. This makes manual editing a bit off, especially because "[]" currently has no auto indentation.

3. Readjustment of cursor when keyboard popups is still covered by symbol input bar.

4. The symbol input bar is a little limited (understandable) and in need of some more frequently used characters such as (), =, +, -, ... Maybe we could give each key the "hold-and-slide" to select more related keys option, just like what our onscreen keyboard has.

5. In rare occasions, symbol input bar is not responding to touch. (working again after reopening file)

**Comment by dhaigit** [comment]
1. Text with 1-level {} now gets pasted correctly, but nested {} still have indentation problems.

2. It's "kinda" better but it still doesn't work as well as it does in a java file.

Examples:
-backspace does not remove the whole indentation as in java.
-try: open a json file; move cursor to the right of the first "["; tap return key; tab backspace key twice; tap tab key; type "abc"; tap return key;

3. Text now gets readjusted above symbol input bar. However, text is still blocked by the horizontal scrollbar for a few seconds. That horizontal bar appears in all file types except for json. We still need that bar at the right times, just not after a keyboard popup.

4. I've noticed there's one changed key. I wish that key were the one I needed, or at least customizable. This is because I'm using a keyboard with a layout that does not have a forward slash "/". A key slot in that symbol input bar that pops up some customizable characters would have been so nice.

5. For this rare issue, I just don't know how to reproduce it (or if it's fixed.) Should it be put in a new issue?

**Comment by tyron12233** [comment]
For the custom symbols, its already planned and will have support for configurable actions such as pasting custom snippets, showing the popup window etc

**Comment by dhaigit** [comment]
1. Nested {} still are not working correctly. I tested with the following example by cutting and pasting the nested for loop inside the i-for-loop:

for (int i = 0; i < 10; i++) {

}

for (int j = 0; j < 10; j++) {
    for (int k = 0; k < 10; k++) {
        //stuff
    }
}

2. Fixed

3. Horizontal scrollbar is still blocking text for a few seconds right after a keyboard popup.

To reproduce:
-make sure a java page has enough text to fill the whole screen and the keyboard is not shown.
-tap on the very bottom line of text.
-keyboard pops up with horizontal scroll bar blocking the current line of text
-horizontal scroll bar dismisses after a few seconds. (but we need to see the line right away after the keyboard pops up)

4. ...

5. ...

**Comment by tyron12233** [comment]
for 3, i might need to forward the issue to sora editor's github repo as the text adjustment is handled there. it may be possible to override the method that handles this but I haven't looked at it yet

**Comment by dhaigit** [comment]
I will retest 1 for a confirmation after the next release as I don't try to build it myself every time.

For 5, haven't I seen somewhere in one of your changelogs that it was fix? (or was it a different issue... I'm not sure.)

**Comment by dhaigit** [comment]
1. Fixed
2. Fixed
3. Pending
4. Planned
5. Not sure

**Comment by tyron12233** [comment]
For #3, Its better to forward the issue to the editor library's repository as I dont see a way of intercepting the calls when the editor has been resized without recreating the whole logic.

**Comment by dhaigit** [comment]
[Update]

1. Fixed
2. Fixed
3. Fixed (in editor's library but not updated in CodeAssist, and not tested)
4. Planned
5. Fixed (hopefully)
